compiler/rustc_hir_analysis/src/hir_ty_lowering/mod.rs

```diff
@@ -1629,26 +1825,92 @@ impl<'tcx> dyn HirTyLowerer<'tcx> + '_ {
         debug!(?trait_def_id);
 
         let Some(self_ty) = opt_self_ty else {
-            return self.error_missing_qpath_self_ty(trait_def_id, span, item_segment);
+            let guar = self.error_missing_qpath_self_ty(
+                trait_def_id,
+                span,
+                item_segment,
+                ty::AssocKind::Type,
+            );
+            return Ty::new_error(tcx, guar);
         };
         debug!(?self_ty);
 
+        let (item_def_id, item_args) = self.lower_qpath_shared(
+            span,
+            self_ty,
+            trait_def_id,
+            item_def_id,
+            trait_segment,
+            item_segment,
+        );
+        Ty::new_projection_from_args(tcx, item_def_id, item_args)
+    }
+
+    /// Lower a qualified path to a const.
+    #[instrument(level = "debug", skip_all)]
+    fn lower_qpath_const(
+        &self,
+        span: Span,
+        opt_self_ty: Option<Ty<'tcx>>,
+        item_def_id: DefId,
+        trait_segment: &hir::PathSegment<'tcx>,
+        item_segment: &hir::PathSegment<'tcx>,
+    ) -> Const<'tcx> {
+        let tcx = self.tcx();
+
+        let trait_def_id = tcx.parent(item_def_id);
+        debug!(?trait_def_id);
+
+        let Some(self_ty) = opt_self_ty else {
+            let guar = self.error_missing_qpath_self_ty(
+                trait_def_id,
+                span,
+                item_segment,
+                ty::AssocKind::Const,
+            );
+            return Const::new_error(tcx, guar);
+        };
+        debug!(?self_ty);
```

Comment:
this should all be moved into `lower_qpath_shared`. I think you could make it return a `Result<(DefId, GenericArgsRef<'tcx>), ErrorGuaranteed)>` then either construct a `ConstKind::Unevaluated` or a `ConstKind::Error` (or `TyKind::Alias`/`TyKind::Error`) depending on whether its ok/err

Reply:
^ still needs doing